File: Operations.py
import numpy as np
from FieldsHolder import FieldArray, MAX_DIM
from MeshObject import MeshObject
import time
from LinearSystem import LinearSystem

class PerformanceTimer:
	def __init__(self):
		self.myStartTimer_ = 0.0
		self.myEndTimer_ = 0.0
		self.myEventName_ = None
	def start_timer(self, eventName):
		self.myEventName_ = eventName
		self.myStartTimer_ = time.perf_counter()
	def end_timer(self):
		assert self.myEventName_ != None, "No event was started"
		self.myEndTimer_ = time.perf_counter()
		print(f"For event {self.myEventName_}, timer: {self.myEndTimer_ - self.myStartTimer_:.6f} seconds ")
		self.myEventName_ = None


class LogObject:
	def __init__(self,systemVector):
		self.mySystemVector_ : list[LinearSystem] = systemVector
		self.myNonlinearResiduals_ = {}
	def report_log(self, nonlinIterCounter):
		print(f"For nonlinear iteration = {nonlinIterCounter}")
		for i in range(len(self.mySystemVector_)):
			system = self.mySystemVector_[i]
			residual = np.linalg.norm(system.get_rhs())
			print(f"\tFor system {system.get_name()}: RHS residual = {residual}")
			self.myNonlinearResiduals_[system.name_] = residual
	def get_residuals(self):
		for i in range(len(self.mySystemVector_)):
			system = self.mySystemVector_[i]
			residual = np.linalg.norm(system.get_rhs())
			self.myNonlinearResiduals_[system.name_] = residual
		return self.myNonlinearResiduals_




class CFLTimeStepCompute:
    def __init__(self, mesh_object, CFL):
        self.myMeshObject_ : MeshObject = mesh_object
        self.CFL_ = CFL
    def compute_time_step(self):
        """Assume a unit velocity at top boundary for now, and uniform cell spacing in x and y"""
        xRange = self.myMeshObject_.get_x_range()
        yRange = self.myMeshObject_.get_y_range()
        numCellsX, numCellsY = self.myMeshObject_.get_cell_count()
        minDx = (xRange[1] - xRange[0])/numCellsX
        minDy = (yRange[1] - yRange[0])/numCellsY
        minCellSpacing = min(minDx, minDy)
        maxVelocity = 1.0
        min_time_step = self.CFL_ * minCellSpacing / maxVelocity
        return min_time_step

class ComputeInteriorMassFlux:
    def __init__(self, mesh_object, velocity_field_array, pressure_field_array, cell_pressure_gradient_field_array, dt):
        self.myMeshObject_ : MeshObject = mesh_object
        self.velocity_field_array_ : FieldArray = velocity_field_array
        self.pressure_field_array_ : FieldArray = pressure_field_array
        self.cell_pressure_gradient_field_array_ : FieldArray = cell_pressure_gradient_field_array # assume cell gradient has already been computed. . .
        self.dt_ = dt
        self.scaling_ = self.dt_
    def correct_mass_flux(self,dP : FieldArray):
        for face in self.myMeshObject_.get_internal_faces():
            leftCellID = face.get_left_cell()
            rightCellID = face.get_right_cell()


            # Get the pressure valuse for the left and right cells
            left_dP = dP.get_data()[leftCellID]
            right_dP = dP.get_data()[rightCellID]
            normalVector = face.get_normal_vector()
            cellLeft = self.myMeshObject_.get_cell_by_flat_id(leftCellID)
            cellRight = self.myMeshObject_.get_cell_by_flat_id(rightCellID)
            distance = 0.0
            for i in range(MAX_DIM):
                distance += (cellRight.get_centroid()[i] - cellLeft.get_centroid()[i])*normalVector[i]
            faceGradPressureCorrection_dot_n = (right_dP - left_dP)/distance
            # Store the computed mass flux in the face object (you may want to have a separate field array for this in practice)
            face.massFlux_ -= self.scaling_ * faceGradPressureCorrection_dot_n

    def compute_mass_flux(self):
        """Compute the mass flux across each face in the mesh based on the velocity field and pressure gradient"""
        for face in self.myMeshObject_.get_internal_faces():
            leftCellID = face.get_left_cell()
            rightCellID = face.get_right_cell()
            
            # Get the velocity values for the left and right cells
            leftVelocity_x = self.velocity_field_array_.get_data()[leftCellID*MAX_DIM + 0]
            leftVelocity_y = self.velocity_field_array_.get_data()[leftCellID*MAX_DIM + 1]
            rightVelocity_x = self.velocity_field_array_.get_data()[rightCellID*MAX_DIM + 0]
            rightVelocity_y = self.velocity_field_array_.get_data()[rightCellID*MAX_DIM + 1]

            # Get the pressure gradient values for the left and right cells
            leftGradPressure_x = self.cell_pressure_gradient_field_array_.get_data()[leftCellID*MAX_DIM + 0]
            leftGradPressure_y = self.cell_pressure_gradient_field_array_.get_data()[leftCellID*MAX_DIM + 1]
            rightGradPressure_x = self.cell_pressure_gradient_field_array_.get_data()[rightCellID*MAX_DIM + 0]
            rightGradPressure_y = self.cell_pressure_gradient_field_array_.get_data()[rightCellID*MAX_DIM + 1]

            # Get the pressure valuse for the left and right cells
            leftPressure = self.pressure_field_array_.get_data()[leftCellID]
            rightPressure = self.pressure_field_array_.get_data()[rightCellID]

            # Compute the mass flux contribution for this face based on rhie-chow interpolation.
            normalVector = face.get_normal_vector()
            
            # Average velocity/gradP at the face
            faceVelocity_x = 0.5*(leftVelocity_x + rightVelocity_x)
            faceVelocity_y = 0.5*(leftVelocity_y + rightVelocity_y)
            cellGradPressure_x = 0.5*(leftGradPressure_x + rightGradPressure_x)
            cellGradPressure_y = 0.5*(leftGradPressure_y + rightGradPressure_y)

            cellLeft = self.myMeshObject_.get_cell_by_flat_id(leftCellID)
            cellRight = self.myMeshObject_.get_cell_by_flat_id(rightCellID)
            distance = 0.0
            for i in range(MAX_DIM):
                distance += (cellRight.get_centroid()[i] - cellLeft.get_centroid()[i])*normalVector[i]
            faceGradPressure_dot_n = (rightPressure - leftPressure)/distance
            u_dot_n = faceVelocity_x*normalVector[0] + faceVelocity_y*normalVector[1]
            cellGradP_dot_n = cellGradPressure_x*normalVector[0] + cellGradPressure_y*normalVector[1]
            # Compute mass flux as velocity dot normal times area of the face
            mass_flux = (u_dot_n + self.scaling_ *((cellGradP_dot_n-faceGradPressure_dot_n))) * face.get_area()
            
            # Store the computed mass flux in the face object (you may want to have a separate field array for this in practice)
            face.massFlux_ = mass_flux

class ComputeCellGradient:
    """Class to compute cell gradients for a given field array"""
    def __init__(self, mesh_object, in_field_array, out_field_array):
        self.mesh_object_ : MeshObject = mesh_object
        self.field_array_ : FieldArray = in_field_array
        self.out_field_array_ : FieldArray = out_field_array
    
    def compute_scalar_gradient(self):
        """Compute the gradient of the field array for each cell in the mesh"""
        self.out_field_array_.initialize_constant(0.0)  # Initialize output array to zero
        numComponents = self.out_field_array_.get_num_components()
        assert numComponents == 2, "Output field array must be a vector field with 2 components for 2D gradient of a scalar"
        for face in self.mesh_object_.get_internal_faces():
            left_cell_id = face.get_left_cell()
            right_cell_id = face.get_right_cell()
            
            # Get the field values for the left and right cells
            left_value = self.field_array_.get_data()[left_cell_id]
            right_value = self.field_array_.get_data()[right_cell_id]
            
            # Compute the gradient contribution for this face
            faceValue = 0.5*(right_value + left_value)
            
            # Add the contribution to the output field array for both cells
            areaVector = (face.get_normal_vector()[0]*face.get_area(), face.get_normal_vector()[1]*face.get_area())
            self.out_field_array_.get_data()[left_cell_id*numComponents + 0] += faceValue*areaVector[0]  # x component contribution for left cell
            self.out_field_array_.get_data()[left_cell_id*numComponents + 1] += faceValue*areaVector[1]  # y component contribution for left cell
            self.out_field_array_.get_data()[right_cell_id*numComponents + 0] -= faceValue*areaVector[0]  # Opposite contribution for right cell, inverted normal
            self.out_field_array_.get_data()[right_cell_id*numComponents + 1] -= faceValue*areaVector[1]  # Opposite contribution for right cell, inverted normal

        for face in self.mesh_object_.get_boundary_faces():
            directionFactor = 1.0
            cellID = face.get_left_cell()  # For boundary faces, only one cell contributes
            faceValue = self.field_array_.get_data()[cellID]  # For boundary faces, use the value from the single adjacent cell


            # Add the contribution to the output field array for both cells
            areaVector = (face.get_normal_vector()[0]*face.get_area(), face.get_normal_vector()[1]*face.get_area())
            self.out_field_array_.get_data()[cellID*numComponents + 0] += directionFactor*faceValue*areaVector[0]  # x component contribution for left cell
            self.out_field_array_.get_data()[cellID*numComponents + 1] += directionFactor*faceValue*areaVector[1]  # y component contribution for left cell
        
        # scale cell output by volume of the cell to get average gradient
        for cell in self.mesh_object_.get_cells():
            cell_id = cell.get_flat_id()
            cell_volume = cell.get_volume()
            self.out_field_array_.get_data()[cell_id*numComponents + 0] /= cell_volume
            self.out_field_array_.get_data()[cell_id*numComponents + 1] /= cell_volume

File: CPP_Impl/src/Operations.cpp
#include "Operations.h"
#include <iostream>
#include <algorithm>
#include <cmath>
#include <numeric>
#include <stdexcept>
#include <mpi.h>

namespace Vinci4D {

namespace {
bool isRootRank() {
    int rank = 0;
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    return rank == 0;
}
}

void PerformanceTimer::startTimer(const std::string& eventName) {
    eventName_ = eventName;
    started_ = true;
    startTime_ = std::chrono::high_resolution_clock::now();
}

void PerformanceTimer::endTimer() {
    if (!started_) {
        if (isRootRank()) {
            std::cerr << "Warning: Timer was not started" << std::endl;
        }
        return;
    }
    
    auto endTime = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(endTime - startTime_);
    double seconds = duration.count() / 1e6;
    
    if (isRootRank()) {
        std::cout << "For event " << eventName_ << ", timer: " << seconds << " seconds" << std::endl;
    }
    eventTimings_[eventName_].push_back(seconds);
    
    started_ = false;
    eventName_ = "";
}

void PerformanceTimer::report_timing_statistics() const {
    if (!isRootRank()) {
        return;
    }
    std::cout << "\n=== Timing Statistics ===" << std::endl;
    for (const auto& pair : eventTimings_) {
        const auto& times = pair.second;
        double total = std::accumulate(times.begin(), times.end(), 0.0);
        double average = total / times.size();
        std::cout << "Event: " << pair.first << ", Average Time: " << average << " seconds over " << times.size() << " events." << std::endl;
    }
}

void LogObject::reportLog(int nonlinIterCounter) {
    if (!isRootRank()) {
        return;
    }
    std::cout << "For nonlinear iteration = " << nonlinIterCounter << std::endl;
    
    for (auto* system : systemVector_) {
        double residual = system->getRhsNorm();
        std::cout << "\tFor system " << system->getName() << ": RHS residual = " << residual << std::endl;
        nonlinearResiduals_[system->getName()] = residual;
    }
}

std::map<std::string, double> LogObject::getResiduals() {
    for (auto* system : systemVector_) {
        double residual = system->getRhsNorm();
        nonlinearResiduals_[system->getName()] = residual;
    }
    return nonlinearResiduals_;
}

double CFLTimeStepCompute::computeTimeStep() {
    auto xRange = meshObject_->getXRange();
    auto yRange = meshObject_->getYRange();
    auto cellCount = meshObject_->getCellCount();
    
    double minDx = (xRange[1] - xRange[0]) / cellCount[0];
    double minDy = (yRange[1] - yRange[0]) / cellCount[1];
    double minCellSpacing = std::min(minDx, minDy);
    
    // Assume unit velocity for now
    double maxVelocity = 1.0;
    
    return CFL_ * minCellSpacing / maxVelocity;
}

ComputeInteriorMassFlux::ComputeInteriorMassFlux(
    MeshObject* meshObject, FieldArray* velocityField,
    FieldArray* pressureField, FieldArray* cellPressureGradientField, double dt)
    : meshObject_(meshObject),
      velocityFieldArray_(velocityField),
      pressureFieldArray_(pressureField),
      cellPressureGradientFieldArray_(cellPressureGradientField),
      dt_(dt),
      scaling_(dt) {}

void ComputeInteriorMassFlux::computeMassFlux() {
    // Exchange ghost cells before computing face-based quantities
    velocityFieldArray_->exchangeGhostCells(*meshObject_);
    pressureFieldArray_->exchangeGhostCells(*meshObject_);
    cellPressureGradientFieldArray_->exchangeGhostCells(*meshObject_);
    
    const auto& internalFaces = meshObject_->getInternalFaces();
    auto& velocityData = velocityFieldArray_->getData();
    auto& pressureData = pressureFieldArray_->getData();
    auto& gradPressureData = cellPressureGradientFieldArray_->getData();
    
    int numCells = meshObject_->getNumCells();
    bool isParallel = (meshObject_->getMpiSize() > 1);
    
    for (auto* face : internalFaces) {
        int leftCellID = face->getLeftCell();
        int rightCellID = face->getRightCell();
        
        // Validate cell IDs
        if (leftCellID < 0 || leftCellID >= numCells || rightCellID < 0 || rightCellID >= numCells) {
            throw std::runtime_error("Invalid cell ID in internal face");
        }

        // Only compute faces owned by this rank in parallel
        if (isParallel && !meshObject_->isLocalCell(leftCellID)) {
            continue;
        }

        int localLeftID = isParallel ? meshObject_->globalToLocal(leftCellID) : leftCellID;
        int localRightID = isParallel ? meshObject_->globalToLocal(rightCellID) : rightCellID;
        if (localLeftID < 0 || localRightID < 0) {
            continue;
        }
        
        // Get velocity values
        double leftVelX = velocityData[localLeftID * MAX_DIM + 0];
        double leftVelY = velocityData[localLeftID * MAX_DIM + 1];
        double rightVelX = velocityData[localRightID * MAX_DIM + 0];
        double rightVelY = velocityData[localRightID * MAX_DIM + 1];
        
        // Get pressure gradient values
        double leftGradPX = gradPressureData[localLeftID * MAX_DIM + 0];
        double leftGradPY = gradPressureData[localLeftID * MAX_DIM + 1];
        double rightGradPX = gradPressureData[localRightID * MAX_DIM + 0];
        double rightGradPY = gradPressureData[localRightID * MAX_DIM + 1];

        // Get pressure values
        double leftPressure = pressureData[localLeftID];
        double rightPressure = pressureData[localRightID];
        
        // Average velocities and pressure gradients
        double avgVelX = 0.5 * (leftVelX + rightVelX);
        double avgVelY = 0.5 * (leftVelY + rightVelY);
        double avgGradPX = 0.5 * (leftGradPX + rightGradPX);
        double avgGradPY = 0.5 * (leftGradPY + rightGradPY);

        // Compute face normal distance between cell centroids
        const Cell* leftCell = meshObject_->getCellByFlatId(leftCellID);
        const Cell* rightCell = meshObject_->getCellByFlatId(rightCellID);
        if (!leftCell || !rightCell) {
            continue;
        }
        auto leftCentroid = leftCell->getCentroid();
        auto rightCentroid = rightCell->getCentroid();
        auto normal = face->getNormalVector();
        double distance = 0.0;
        for (int i = 0; i < MAX_DIM; ++i) {
            distance += (rightCentroid[i] - leftCentroid[i]) * normal[i];
        }
        if (std::abs(distance) < 1e-12) {
            continue;
        }

        // Rhie-Chow interpolation: add face pressure gradient correction
        double uDotN = avgVelX * normal[0] + avgVelY * normal[1];
        double cellGradPDotN = avgGradPX * normal[0] + avgGradPY * normal[1];
        double faceGradPDotN = (rightPressure - leftPressure) / distance;
        double massFlux = (uDotN + scaling_ * (cellGradPDotN - faceGradPDotN)) * face->getArea();
        
        // Store in face (direct access like Python version)
        face->massFlux_ = massFlux;
    }
}

void ComputeCellGradient::computeScalarGradient() {
    // Exchange ghost cells before computing gradients
    inField_->exchangeGhostCells(*meshObject_);
    
    outField_->initializeConstant(0.0);
    
    int numComponents = outField_->getNumComponents();
    auto& inData = inField_->getData();
    auto& outData = outField_->getData();
    
    bool isParallel = (meshObject_->getMpiSize() > 1);

    // Interior faces contribution
    for (const auto* face : meshObject_->getInternalFaces()) {
        int leftCellID = face->getLeftCell();
        int rightCellID = face->getRightCell();
        
        if (leftCellID < 0 || rightCellID < 0) continue;
        if (leftCellID >= meshObject_->getNumCells() || rightCellID >= meshObject_->getNumCells()) continue;
        
        int localLeftID = isParallel ? meshObject_->globalToLocal(leftCellID) : leftCellID;
        int localRightID = isParallel ? meshObject_->globalToLocal(rightCellID) : rightCellID;
        if (localLeftID < 0 || localRightID < 0) {
            continue;
        }

        double leftValue = inData[localLeftID];
        double rightValue = inData[localRightID];
        double faceValue = 0.5 * (leftValue + rightValue);
        
        auto normal = face->getNormalVector();
        double area = face->getArea();
        double areaX = normal[0] * area;
        double areaY = normal[1] * area;
        
        // Left cell contribution
        if (!isParallel || meshObject_->isLocalCell(leftCellID)) {
            outData[localLeftID * numComponents + 0] += faceValue * areaX;
            outData[localLeftID * numComponents + 1] += faceValue * areaY;
        }
        
        // Right cell contribution (opposite normal)
        if (!isParallel || meshObject_->isLocalCell(rightCellID)) {
            outData[localRightID * numComponents + 0] -= faceValue * areaX;
            outData[localRightID * numComponents + 1] -= faceValue * areaY;
        }
    }
    
    // Boundary faces contribution
    for (const auto* face : meshObject_->getBoundaryFaces()) {
        int cellID = face->getLeftCell();
        if (cellID < 0 || cellID >= meshObject_->getNumCells()) continue;
        if (isParallel && !meshObject_->isLocalCell(cellID)) continue;

        int localCellID = isParallel ? meshObject_->globalToLocal(cellID) : cellID;
        if (localCellID < 0) continue;
        
        double faceValue = inData[localCellID];
        auto normal = face->getNormalVector();
        double area = face->getArea();
        
        outData[localCellID * numComponents + 0] += faceValue * normal[0] * area;
        outData[localCellID * numComponents + 1] += faceValue * normal[1] * area;
    }
    
    // Scale by cell volume
    const auto& cells = isParallel ? meshObject_->getLocalCells() : meshObject_->getCells();
    for (const auto& cell : cells) {
        int localCellID = isParallel ? cell.getLocalId() : cell.getFlatId();
        double cellVolume = cell.getVolume();
        
        outData[localCellID * numComponents + 0] /= cellVolume;
        outData[localCellID * numComponents + 1] /= cellVolume;
    }
}

} // namespace Vinci4D
